book/6-export.md

```diff
@@ -0,0 +1,118 @@
+---
+jupytext:
+  formats: md:myst
+  text_representation:
+    extension: .md
+    format_name: myst
+    format_version: 0.13
+    jupytext_version: 1.16.4
+kernelspec:
+  display_name: Python 3 (ipykernel)
+  language: python
+  name: python3
+---
+
+# Exporting
+
++++
+
+In this chapter we'll look at how to convert sub-volumes of 3D zarr images to other file formats.
+The use case for this is allow users to extract sub-volumes of large 3D images into a preferred file format, perform some analysis/processing on the sub-volume using their preferred tools and then copy the results back in-place to the original zarr image.
+
+## Converting Zarr sub-volumes to tiff
+
+To begin this chapter, we'll look at how to convert Zarr sub-volumes to tiff.
+
+To start with we'll load our heart sample dataset, as a zarr array and then extract a smaller volume as a NumPy array.
+
+Note: the heart sample dataset is only 200MB, so the entire dataset could be extracted into a NumPy array (and therefore loaded into memory) if required.
+If you are working with larger datasets, you will need to keep the extracted volume small enough to fit into memory, or extract it as a series of smaller chunks.
+If you are using an in-memory Zarr store, then the entire dataset will be loaded into memory regardless of the size of the sub-volume you extract.
+
+```{code-cell} ipython3
+from pathlib import Path
+import zarr
+import numpy as np
+
+from data_helpers import load_heart_data, plot_slice
+zarr_array = load_heart_data(array_type='zarr')
+original_array = zarr_array[:, :, :].copy()  # Copy with original data
+changed_array = zarr_array[:, :, :].copy()  # Copy for processed data
+start_idx = 40
+end_idx = 50
+slice_range = slice(start_idx, end_idx)
+np_array = zarr_array[slice_range, slice_range, slice_range]
+n_bytes = np_array.nbytes  
+n_gb = n_bytes / (10**9)
+# Print store information to show data is not loaded into memory
+print(f"Zarr array store: {zarr_array.store}")
+print(f"Zarr array shape: {zarr_array.shape}")
+print(f"Zarr array dtype: {zarr_array.dtype}")
+print(f"Zarr array chunks: {zarr_array.chunks}")
+print(f"Sub-volume size: {n_gb:.3f} GB")
+print(f"Only the requested slice is loaded into memory as a NumPy array")
+```
+
+The next step is to write this NumPy array to a tiff file.
+We'll use the `imageio` library to do this.
+
+Processing of the sub-volume can be done using any tool that can read and write tiff files.
+For example, you could use `Fiji`.
+
+The `imageio` library is used to then read the tiff file back into a NumPy array.
+This subvolume array can then be copied back in place to the original Zarr array.
+
+```{code-cell} ipython3
+import imageio.v3 as iio
+# iio.imwrite("image_file.tiff", np_array, plugin='tifffile')
+zarr_small = np_array
```

Comment:
At this point you should break the code cell and put in some actual text explaining that one can now do some processing on the iamge using another tool. Then explain instead of that, for the purposes of the tutorial you'll fake some processing happening on the data.